sn_interface/src/network_knowledge/sections_dag.rs

```diff
@@ -0,0 +1,1020 @@
+            self.dag_root.insert(key);
+            BTreeSet::new()
+        } else {
+            let hash = self
+                .hashes
+                .get(parent_key)
+                .ok_or(Error::KeyNotFound(*parent_key))?;
+            BTreeSet::from([*hash])
+        };
+
+        let info = SectionInfo {
+            key,
+            sig: signature,
+        };
+        // dag's children arg = parent
+        let child = self.dag.write(info, parent);
+        self.hashes.insert(key, child.hash());
+        self.dag.apply(child);
+
+        Ok(())
+    }
+
+    /// Insert new key into the DAG. `parent_key` must exist in the DAG and must validate
+    /// `signature`, otherwise error is returned.
+    fn insert_node(&mut self, parent_key: &bls::PublicKey, node: Node<SectionInfo>) -> Result<()> {
+        self.insert(parent_key, node.value.key, node.value.sig)
+    }
+
+    /// Get a partial `SectionsDAG` from the `from` key to the `to` key
+    /// Returns `Error::KeyNotFound` if the `to` key is not present in the DAG;
+    /// Returns `Error::InvalidBranch` if the `from` key is not found or is not a direct ancestor of the `to` key
+    pub fn partial_dag(&self, from: &bls::PublicKey, to: &bls::PublicKey) -> Result<Self> {
+        // start from the "to" key (bottom of the tree) and traverse to the root
+        let mut crdt_ops: Vec<Node<SectionInfo>> = Vec::new();
+        // if "to" key is genesis, returns error
+        let mut node = self.get_node(to)?;
+        loop {
+            if node.value.key == *from {
+                // we have reached the end
+                break;
+            }
+            // we don't insert the `from` node
+            crdt_ops.push(node.clone());
+            match self.parent_node(node.hash()) {
+                Some(parent) => node = parent,
+                // reached root of the dag
+                None => {
+                    // after reaching dag_root, return the partial_dag if from == genesis
+                    if *from == self.genesis_key {
+                        break;
+                    }
+                    // from key was apparently not a direct ancestor
+                    return Err(Error::InvalidBranch);
+                }
+            }
+        }
+        // create the SectionsDAG and apply the CRDT ops
+        let mut dag = SectionsDAG::new(*from);
+        let mut parent = *from;
+        for node in crdt_ops.into_iter().rev() {
+            let key = node.value.key;
+            dag.insert_node(&parent, node)?;
+            parent = key;
+        }
+        Ok(dag)
+    }
+
+    /// Update our current `SectionsDAG` with the keys from another `SectionsDAG`
+    /// Returns `Error::InvalidSignature` if the provided DAG fails signature verification
+    /// Returns `Error::KeyNotFound` if the genesis_key of either of the DAGs is not present in the
+    /// other
+    pub fn join(&mut self, mut other: Self) -> Result<()> {
```

Comment:
I think `merge` is more clear
```suggestion
    pub fn merge(&mut self, mut other: Self) -> Result<()> {
```